### ISSUE TITLE
[New Service]: AWS User Notifications

### ISSUE THREAD

**Issue by F-Xu** [hashicorp/terraform-provider-aws#34969]
### Description

AWS announced User Notifications general availability on 3rd May 2023.

### Requested Resource(s) and/or Data Source(s)

aws_usernotification_configuration
aws_usernotification_deliverychannel
aws_usernotification_hub

### References

https://aws.amazon.com/about-aws/whats-new/2023/05/aws-user-notifications-available/
https://docs.aws.amazon.com/notifications/latest/userguide/what-is-service.html

### Would you like to implement a fix?

None

**Comment by colonmelvin** [comment]
Per [AWS documentation](https://aws.amazon.com/notifications/faqs/) there is no API for User Notifications. 

_Does AWS User Notifications provide users with APIs to access notifications?_
No, AWS User Notifications does not provide programmatic access to notifications.

**Comment by iwamot** [comment]
AWS User Notifications API is now available.
https://docs.aws.amazon.com/notifications/latest/APIReference/Welcome.html

**Comment by jatin-mehrotra-colorkrew** [comment]
https://aws.amazon.com/about-aws/whats-new/2024/11/aws-user-notifications-sdk/

SDK is available which also means API is also available

**Comment by DanHyland** [comment]
Bump now that the API and SDK are now available. Is there an ETA when we can have TF resources to manage AWS User Notifications?

**Comment by marcinbelczewski** [comment]
@srgoni are willing to contribute the design proposed above or possibly you'd welcome some assistance? I could contribute as well.

**Comment by marcinbelczewski** [comment]
Thanks for comment @srgoni. I will try to take a stab and if I'm successful I will create a draft PR. Cheers!

**Comment by marcinbelczewski** [comment]
Here are my comments @srgoni:

Ad 1: I believe it absolutely makes sense to reference AWS docs here. Good idea.

Ad 2: I decided to skip those after struggling with `status_summary` for a moment. "Flexed" response would need to land into nested attributes which are currently supported in protocol ver 6 while this provider is still on 5. So I'd need to make this a little less pretty and since those aren't really used for anything meaningful (I reckon) I decided to skip those.

Ad 3: Current implementation already waits for the deregistration of the hub. I believe what takes place here, is parallel invocation of destruction and creation starting almost instanteanously resulting in conflict. The way I'd migrate the code from one to another would be:

```terraform
resource "aws_notifications_notification_hub" "test" {
  for_each = toset(["us-west-1"])
  notification_hub_region = each.value
}

moved {
  from = aws_notifications_notification_hub.us-west-1
  to   = aws_notifications_notification_hub.test["us-west-1"]
}
```

Please let me know if that makes sense.

Ad 4. Yup - three hubs per account is a current limit. Not sure how to fail quickly on this though since registration can be happening from multiple different processes (some not even Terraform) and there is no locking mechanism to serialize access. I will think about this some more tomorrow and will post here what I came up with. Stay tuned ;-)

Ad 5. This is tricky for sure. Singleton resource with idempotent API. Not sure there's good support for those in Terraform so definitely something that could be mentioned in documentation as you suggest. 

Cheers!

**Comment by marcinbelczewski** [comment]
Hi all. I have just received the response from AWS regarding the erroneous behavior of parallel Notification Hub registrations. Here's the response quoted in full:

_I've received an update from the internal service team in regards to the behavior and findings I had relayed to them via the internal escalation._

_As per their update they confirmed via internal logging that the parallel calls leading to higher then intended hub registration is unintentional and occurs due to the shortness in time between the registration requests initiated in parallel leading to a bypass in the throttling system that should have triggered at 3 hubs. As such the internal team is actively working on a fix for this unintentional behavior._

_As for the queuing behavior in relation to the registration and de registration actions, the reason this behavior is seen is due to a side-effect of the List API restricting the max results to 3 as there should only be 3 hubs. Only 3 of the 5 hubs are displayed accordingly and de-registering 2 of those hubs pulled in the remaining two into the list results._

_With the information above in mind we appreciate your understanding and bringing this behavior to our attention. As for now the internal service team recommends calling the API serially rather then via parallel calls. I informed your TAM, Michael Dobson, as to the current status of this case and can assist in tracking the status of the fix post case. Due to the nature of supports limited direct visibility into service developments, I am unable to provide an exact time of when the fix will be completed by the service team._

So I reckon for now we shouldn't be fixing anything in the provider when it comes to those particular problems and wait for AWS to fix things on their side. Please let me know if that makes sense

**Comment by wernerdiers** [comment]
Hi all, AWS has release the optional setting to manage User Notifications at the Organization level. Now one can optionally configure a notification on a specific OU.

Docs
- https://docs.aws.amazon.com/notifications/latest/userguide/managing-org-notifications.html

API
- https://docs.aws.amazon.com/notifications/latest/APIReference/API_AssociateOrganizationalUnit.html

aws cli
- https://docs.aws.amazon.com/cli/latest/reference/notifications/associate-organizational-unit.html

Perhaps this setting could be added to this feature too?

**Comment by madhavvishnubhatta** [comment]
Created a separate issue for Managed Notifications - https://github.com/hashicorp/terraform-provider-aws/issues/45133

**Comment by madhavvishnubhatta** [comment]
I have started working on Organizational Unit association.
